File: go/validate.go
package protocol

import (
	"encoding/json"
	"errors"
	"fmt"
	"strconv"
	"strings"

	"github.com/nbd-wtf/go-nostr"
)

// Verdict is a validation outcome plus a stable reason code (spec/envelope.md §2).
type Verdict struct {
	Accept bool
	Code   string
}

// EventModelError describes a schema validation failure for a custom kind.
// Compatible with the former eventmodel.EventModelError so the relay's existing
// error handling keeps working after the migration.
type EventModelError struct {
	Kind   int
	Reason string
}

func (e *EventModelError) Error() string {
	return fmt.Sprintf("invalid kind %d event: %s", e.Kind, e.Reason)
}

// AsEventModelError converts a Verdict into an EventModelError (or nil when
// accepted). The reason is the verdict code itself.
func (v Verdict) AsEventModelError(kind int) error {
	if v.Accept {
		return nil
	}
	return &EventModelError{Kind: kind, Reason: v.Code}
}

// Validate applies the event-protocol rules in order and returns the first
// failure. Non-custom / non-corpus kinds are accepted (the relay treats
// standard NIP kinds as opaque). The result codes for the corpus kinds match
// spec/envelope.md §2 exactly.
func Validate(event *nostr.Event) Verdict {
	if event == nil {
		return Verdict{Accept: false, Code: "content-not-json"}
	}
	kind := event.Kind
	body, contentOK := parseContent(event)

	if JSONContentKinds[kind] && !contentOK {
		return Verdict{Accept: false, Code: "content-not-json"}
	}

	d := tagValue(event, "d")
	if AddressableKinds[kind] {
		if d == "" {
			return Verdict{Accept: false, Code: "missing-d"}
		}
		if Hex64DKinds[kind] && !isHex64(d) {
			return Verdict{Accept: false, Code: "d-not-hex64"}
		}
	}

	if schema, ok := body["schema"]; ok {
		if n, ok := schema.(float64); !ok || n < 1 || n != float64(int64(n)) {
			return Verdict{Accept: false, Code: "schema-invalid"}
		}
	}
	if revision, ok := body["revision"]; ok {
		if n, ok := revision.(float64); !ok || n < 0 || n != float64(int64(n)) {
			return Verdict{Accept: false, Code: "revision-invalid"}
		}
	}
	if subject, ok := body["subject"]; ok && d != "" {
		if s, ok := subject.(string); !ok || s != d {
			return Verdict{Accept: false, Code: "subject-mismatch"}
		}
	}

	switch kind {
	case KindCapability:
		if t, ok := body["type"]; !ok {
			return Verdict{Accept: false, Code: "31100:missing-type"}
		} else if s, ok := t.(string); !ok || s == "" {
			return Verdict{Accept: false, Code: "31100:missing-type"}
		}
		if scopes, ok := body["scopes"]; ok {
			list, ok := scopes.([]interface{})
			if !ok {
				return Verdict{Accept: false, Code: "31100:scopes-not-array"}
			}
			for _, s := range list {
				if _, ok := s.(string); !ok {
					return Verdict{Accept: false, Code: "31100:scopes-not-array"}
				}
			}
		}
	case KindTrustPolicy:
		if _, ok := body["schema"]; !ok {
			return Verdict{Accept: false, Code: "31101:missing-schema"}
		}
	case KindIdentityDefinition:
		if enabled, ok := body["enabled"]; ok {
			if _, ok := enabled.(bool); !ok {
				return Verdict{Accept: false, Code: "31102:enabled-not-bool"}
			}
		} else {
			enabled = true
		}
		if admin, ok := body["admin"]; ok {
			if _, ok := admin.(bool); !ok {
				return Verdict{Accept: false, Code: "31102:admin-not-bool"}
			}
		}
		username, _ := body["username"].(string)
		if username == "" && body["enabled"] != false {
			return Verdict{Accept: false, Code: "31102:missing-username"}
		}
		signerType, _ := body["signer_type"].(string)
		if signerType != "" && !SignerTypes[signerType] {
			return Verdict{Accept: false, Code: "31102:bad-signer-type"}
		}
	case KindDelegation:
		p := tagValue(event, "p")
		server := tagValue(event, "server")
		expiry := tagValue(event, "expiry")
		if !isHex64(p) || !isHex64(server) {
			return Verdict{Accept: false, Code: "27236:bad-tags"}
		}
		if expiry == "" {
			return Verdict{Accept: false, Code: "27236:bad-expiry"}
		}
		if n, err := strconv.ParseInt(expiry, 10, 64); err != nil || n <= 0 {
			return Verdict{Accept: false, Code: "27236:bad-expiry"}
		}
		if len(tagsNamed(event, "scope")) == 0 {
			return Verdict{Accept: false, Code: "27236:missing-scope"}
		}
	case KindDelegationRevocation:
		e := tagValue(event, "e")
		if !isHex64(e) {
			return Verdict{Accept: false, Code: "27237:missing-e"}
		}
	case KindPermissionSet:
		for _, p := range tagsNamed(event, "p") {
			if len(p) < 2 || !isHex64(p[1]) {
				return Verdict{Accept: false, Code: "30000:bad-p"}
			}
		}
	case KindMuteList:
		for _, p := range tagsNamed(event, "p") {
			if len(p) < 2 || !isHex64(p[1]) {
				return Verdict{Accept: false, Code: "10000:bad-p"}
			}
		}
	case KindRelayList:
		relays := []string{}
		for _, r := range tagsNamed(event, "r") {
			if len(r) >= 2 {
				relays = append(relays, r[1])
			}
		}
		if len(relays) == 0 {
			return Verdict{Accept: false, Code: "10002:bad-r"}
		}
		for _, u := range relays {
			if !isWebSocketURL(u) {
				return Verdict{Accept: false, Code: "10002:bad-r"}
			}
		}
	case KindOperationRequest:
		var body2 struct {
			Tool string `json:"tool"`
		}
		if err := json.Unmarshal([]byte(event.Content), &body2); err != nil {
			return Verdict{Accept: false, Code: "content-not-json"}
		}
		if strings.TrimSpace(body2.Tool) == "" {
			return Verdict{Accept: false, Code: "operation-request-missing-tool"}
		}
	case KindOperationApproval, KindOperationRejection, KindExecutionStarted, KindExecutionResult, KindExecutionProgress:
		e := tagValue(event, "e")
		if e == "" {
			return Verdict{Accept: false, Code: "chain-step-missing-e"}
		}
		if !contentOK && event.Content != "" {
			return Verdict{Accept: false, Code: "content-not-json"}
		}
	case KindSystemEvent, KindServiceEvent, KindBackupEvent, KindSecurityEvent, KindStateBundle:
		if event.Content != "" && !contentOK {
			return Verdict{Accept: false, Code: "content-not-json"}
		}
	case KindBuildAttestation:
		if d == "" {
			return Verdict{Accept: false, Code: "missing-d"}
		}
		if event.Content != "" && !contentOK {
			return Verdict{Accept: false, Code: "content-not-json"}
		}
	}

	return Verdict{Accept: true}
}

// ValidateError returns an error for a rejected event (nil when accepted or
// non-custom). Compatible with the former eventmodel.Validate contract.
func ValidateError(event *nostr.Event) error {
	if event == nil || !IsCustomKind(event.Kind) {
		return nil
	}
	return Validate(event).AsEventModelError(event.Kind)
}

func parseContent(event *nostr.Event) (map[string]interface{}, bool) {
	if event.Content == "" {
		return map[string]interface{}{}, true
	}
	var body map[string]interface{}
	if err := json.Unmarshal([]byte(event.Content), &body); err != nil {
		return nil, false
	}
	return body, true
}

func tagValue(event *nostr.Event, name string) string {
	tag := event.Tags.Find(name)
	if tag == nil || len(tag) < 2 {
		return ""
	}
	return tag[1]
}

func tagsNamed(event *nostr.Event, name string) [][]string {
	out := [][]string{}
	for _, tag := range event.Tags {
		if len(tag) > 0 && tag[0] == name {
			out = append(out, tag)
		}
	}
	return out
}

func isHex64(s string) bool {
	if len(s) != 64 {
		return false
	}
	for _, c := range s {
		if !((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')) {
			return false
		}
	}
	return true
}

func isWebSocketURL(s string) bool {
	lower := strings.ToLower(s)
	return strings.HasPrefix(lower, "ws://") || strings.HasPrefix(lower, "wss://")
}

// ReasonCode maps an error (EventModelError) to the stable corpus vocabulary.
// Returns "schema-invalid" for unexpected reasons.
func ReasonCode(err error) string {
	var modelErr *EventModelError
	if !errors.As(err, &modelErr) {
		return "schema-invalid"
	}
	code := modelErr.Reason
	if code == "" {
		return "schema-invalid"
	}
	return code
}

// NoticeBody is the optional convention carried by system/service/backup/
// security notices: a free-form "class", a "severity" (SeverityInfo/Warning/
// Critical) and a human-readable "summary". All three are optional.
type NoticeBody struct {
	Class    string `json:"class"`
	Severity string `json:"severity"`
	Summary  string `json:"summary"`
}

// Notice extracts the class/severity/summary convention from a notice event.
// ok is false when the content is empty or not shaped like a notice body
// (callers should fall back to kind-derived defaults).
func Notice(event *nostr.Event) (class, severity, summary string, ok bool) {
	if event == nil || event.Content == "" {
		return "", "", "", false
	}
	var body NoticeBody
	if err := json.Unmarshal([]byte(event.Content), &body); err != nil {
		return "", "", "", false
	}
	return body.Class, body.Severity, body.Summary, true
}

File: python/src/nostrhost_protocol/validate.py
"""Event validation implementing ``spec/envelope.md`` exactly.

Port of the reference implementation (formerly ``tools/event_protocol.py``).
Rules are applied in the documented order; the first failing rule wins and its
stable code is returned.
"""

from __future__ import annotations

import json
import re
from typing import Any

from .kinds import ADDRESSABLE_KINDS, HEX64_D_KINDS, JSON_CONTENT_KINDS, SIGNER_TYPES
from .verdict import Verdict

_HEX64 = re.compile(r"^[0-9a-fA-F]{64}$")
_WS_URL = re.compile(r"^wss?://", re.IGNORECASE)


def is_hex64(value: str) -> bool:
    return bool(_HEX64.match(value))


def _tag(event: dict, name: str) -> list[str] | None:
    for tag in event.get("tags", []):
        if tag and tag[0] == name:
            return tag
    return None


def _tags(event: dict, name: str) -> list[list[str]]:
    return [tag for tag in event.get("tags", []) if tag and tag[0] == name]


def _parse_content(event: dict) -> tuple[object, bool]:
    """Return (parsed, ok). Empty content parses to {} with ok True."""
    raw = event.get("content") or ""
    if raw == "":
        return {}, True
    try:
        return json.loads(raw), True
    except (ValueError, TypeError):
        return None, False


def validate_event(event: dict) -> Verdict:
    """Apply the event-protocol rules in order; return the first failure."""
    kind = int(event.get("kind", 0))
    content, content_ok = _parse_content(event)

    if kind in JSON_CONTENT_KINDS and not content_ok:
        return Verdict(False, "content-not-json")
    body = content if isinstance(content, dict) else {}

    d = _tag(event, "d")
    if kind in ADDRESSABLE_KINDS:
        if d is None or len(d) < 2 or d[1] == "":
            return Verdict(False, "missing-d")
        if kind in HEX64_D_KINDS and not is_hex64(d[1]):
            return Verdict(False, "d-not-hex64")

    if "schema" in body:
        schema = body["schema"]
        if not isinstance(schema, int) or isinstance(schema, bool) or schema < 1:
            return Verdict(False, "schema-invalid")
    if "revision" in body:
        revision = body["revision"]
        if not isinstance(revision, int) or isinstance(revision, bool) or revision < 0:
            return Verdict(False, "revision-invalid")
    if "subject" in body and d is not None and len(d) >= 2 and body["subject"] != d[1]:
        return Verdict(False, "subject-mismatch")

    if kind == 31100:
        if not isinstance(body.get("type"), str) or not body["type"]:
            return Verdict(False, "31100:missing-type")
        scopes = body.get("scopes")
        if scopes is not None and (
            not isinstance(scopes, list) or any(not isinstance(s, str) for s in scopes)
        ):
            return Verdict(False, "31100:scopes-not-array")
    elif kind == 31101:
        if "schema" not in body:
            return Verdict(False, "31101:missing-schema")
    elif kind == 31102:
        enabled = body.get("enabled", True)
        if not isinstance(enabled, bool):
            return Verdict(False, "31102:enabled-not-bool")
        if "admin" in body and not isinstance(body["admin"], bool):
            return Verdict(False, "31102:admin-not-bool")
        username = body.get("username")
        if enabled and (not isinstance(username, str) or not username.strip()):
            return Verdict(False, "31102:missing-username")
        signer_type = body.get("signer_type")
        if signer_type not in (None, "") and signer_type not in SIGNER_TYPES:
            return Verdict(False, "31102:bad-signer-type")
    elif kind == 27236:
        p = _tag(event, "p")
        server = _tag(event, "server")
        if (
            p is None
            or len(p) < 2
            or not is_hex64(p[1])
            or server is None
            or len(server) < 2
            or not is_hex64(server[1])
        ):
            return Verdict(False, "27236:bad-tags")
        expiry = _tag(event, "expiry")
        if expiry is None or len(expiry) < 2:
            return Verdict(False, "27236:bad-expiry")
        try:
            if int(expiry[1]) <= 0:
                raise ValueError
        except (ValueError, TypeError):
            return Verdict(False, "27236:bad-expiry")
        if not _tags(event, "scope"):
            return Verdict(False, "27236:missing-scope")
    elif kind == 27237:
        e = _tag(event, "e")
        if e is None or len(e) < 2 or not is_hex64(e[1]):
            return Verdict(False, "27237:missing-e")
    elif kind == 30000:
        for tag in _tags(event, "p"):
            if len(tag) < 2 or not is_hex64(tag[1]):
                return Verdict(False, "30000:bad-p")
    elif kind == 10000:
        for tag in _tags(event, "p"):
            if len(tag) < 2 or not is_hex64(tag[1]):
                return Verdict(False, "10000:bad-p")
    elif kind == 10002:
        relays = [tag for tag in _tags(event, "r") if len(tag) >= 2]
        if not relays or any(not _WS_URL.match(tag[1]) for tag in relays):
            return Verdict(False, "10002:bad-r")

    return Verdict(True, "")


# Backwards-compatible alias (the reference module used the bare name).
validate = validate_event


def revision_of(event: dict) -> int:
    body, ok = _parse_content(event)
    if ok and isinstance(body, dict) and isinstance(body.get("revision"), int):
        return body["revision"]
    return 0


def declares_revision(event: dict) -> bool:
    body, ok = _parse_content(event)
    return ok and isinstance(body, dict) and "revision" in body


def _content_key(event: dict) -> str:
    """Canonical comparison key for 'would these fold to the same fact?'."""
    body, ok = _parse_content(event)
    if not ok or not isinstance(body, dict):
        return ""
    relevant = {k: v for k, v in body.items() if k not in ("revision", "updated_by", "reason")}
    return json.dumps(relevant, sort_keys=True, separators=(",", ":"))


def fold_events(kind: int, events: list[dict]) -> Any:
    from .fold import FoldResult, fold

    return fold(kind, events)